driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -0,0 +1,679 @@
+    }
+
+    @Nullable
+    private OidcRequestCallback getRequestCallback() {
+        return getMongoCredentialWithCache()
+                .getCredential()
+                .getMechanismProperty(REQUEST_TOKEN_CALLBACK_KEY, null);
+    }
+
+    @Override
+    public void reauthenticate(final InternalConnection connection) {
+        // method must only be called after original handshake:
+        assertTrue(connection.opened());
+        authLock(connection, connection.getDescription());
+    }
+
+    @Override
+    public void authenticate(final InternalConnection connection, final ConnectionDescription connectionDescription) {
+        // method must only be called during original handshake:
+        assertFalse(connection.opened());
+        // this method "wraps" the default authentication method in custom OIDC retry logic
+        String accessToken = getValidCachedAccessToken();
+        if (accessToken != null) {
+            try {
+                authenticateUsing(connection, connectionDescription, (challenge) -> prepareTokenAsJwt(accessToken));
+            } catch (MongoSecurityException e) {
+                if (triggersRetry(e)) {
+                    authLock(connection, connectionDescription);
+                } else {
+                    throw e;
+                }
+            }
+        } else {
+            authLock(connection, connectionDescription);
+        }
+    }
+
+    private static boolean triggersRetry(@Nullable final Throwable t) {
+        if (t instanceof MongoSecurityException) {
+            MongoSecurityException e = (MongoSecurityException) t;
+            Throwable cause = e.getCause();
+            if (cause instanceof MongoCommandException) {
+                MongoCommandException commandCause = (MongoCommandException) cause;
+                return commandCause.getErrorCode() == 18;
+            }
+        }
+        return false;
+    }
+
+    private void authenticateUsing(
+            final InternalConnection connection,
+            final ConnectionDescription connectionDescription,
+            final Function<byte[], byte[]> evaluateChallengeFunction) {
+        this.evaluateChallengeFunction = evaluateChallengeFunction;
+        super.authenticate(connection, connectionDescription);
+    }
+
+    private void authLock(final InternalConnection connection, final ConnectionDescription connectionDescription) {
+        fallbackState = FallbackState.INITIAL;
+        Locks.withLock(getMongoCredentialWithCache().getOidcLock(), () -> {
+            while (true) {
+                try {
+                    authenticateUsing(connection, connectionDescription, (challenge) -> evaluate(challenge));
+                    break;
+                } catch (MongoSecurityException e) {
+                    boolean shouldRetry = triggersRetry(e) && shouldRetryHandler();
+                    if (!shouldRetry) {
+                        throw e;
+                    }
```

Comment:
The async counterpart does not extract `shouldRetry`. Let's remove the variable here to make the code more similar.

```suggestion
                    if (!(triggersRetry(e) && shouldRetryHandler())) {
                        throw e;
                    }
```